src/controller/proxy/controller.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/controller/artifact"
+	"github.com/goharbor/harbor/src/controller/blob"
+	"github.com/goharbor/harbor/src/lib"
+	"github.com/goharbor/harbor/src/lib/errors"
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/replication/model"
+	"github.com/goharbor/harbor/src/replication/registry"
+	"github.com/opencontainers/go-digest"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+const (
+	maxWait             = 10
+	manifestListWaitSec = 1800
+	sleepIntervalSec    = 20
+)
+
+var (
+	// Ctl is a global proxy controller instance
+	Ctl = NewController()
+)
+
+// Controller defines the operations related with pull through proxy
+type Controller interface {
+	// UseLocalManifest check if the manifest should use local
+	UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool
+	// UseLocalBlob check if the blob should use local
+	UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool
+	// ProxyBlob proxy the blob request to the target server
+	ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error
+	// ProxyManifest proxy the manifest to the target server
+	ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error
+}
+
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	mu          sync.Mutex
+	inflight    map[string]interface{}
+	artifactCtl artifact.Controller
+	local       LocalInterface
+}
+
+func (c *controller) isProxyReady(p *models.Project) bool {
+	if p.RegistryID < 1 {
+		return false
+	}
+	reg, err := c.registryMgr.Get(p.RegistryID)
+	if err != nil {
+		log.Errorf("failed to get registry, error:%v", err)
+		return false
+	}
+	return reg.Status == model.Healthy
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	if len(string(art.Digest)) > 0 {
+		exist, err := c.local.BlobExist(ctx, art.Digest)
+		if err == nil && exist {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	exist, err := c.local.BlobExist(ctx, digest)
+	if err != nil {
+		return false
+	}
+	return exist
+}
+
+func (c *controller) ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error {
+	var man distribution.Manifest
+	var err error
+	ref := string(art.Digest)
+	if len(ref) == 0 {
+		ref = art.Tag
+	}
+	if len(ref) == 0 {
+		ref = "latest"
+	}
+	man, err = r.Manifest(repo, ref)
+	if err != nil {
+		if errors.IsNotFoundErr(err) {
+			go func() {
+				c.local.DeleteManifest(ctx, repo, string(art.Tag))
```

Comment:
This isn't needed as the `ProxyManifest` is called only when the manifest isn't in local